### PR TITLE
sds: don't throw exceptions in initialize()

### LINKED ISSUE #10976: Uncaught exception when using filebased xDS and file doesn't exist
When the SDS config points to a file that doesn't exist, an uncaught exception is thrown resulting in `std::terminate`:

```
#0  0x00007ffff7110428 in raise () from /lib/x86_64-linux-gnu/libc.so.6
#1  0x00007ffff711202a in abort () from /lib/x86_64-linux-gnu/libc.so.6
#2  0x000055555a43b099 in Envoy::TerminateHandler::logOnTerminate() const::$_0::operator()() const (this=0x55555a43afd0 <Envoy::TerminateHandler::logOnTerminate() const::$_0::__invoke()>) at external/envoy/source/exe/terminate_handler.cc:15
#3  0x000055555a43afd9 in Envoy::TerminateHandler::logOnTerminate() const::$_0::__invoke() () at external/envoy/source/exe/terminate_handler.cc:12
#4  0x000055555b054806 in __cxxabiv1::__terminate(void (*)()) ()
#5  0x000055555b054851 in std::terminate() ()
#6  0x000055555b060b04 in __cxa_throw ()
#7  0x000055555a075f8e in Envoy::Config::Utility::checkFilesystemSubscriptionBackingPath (path=..., api=...) at external/envoy/source/common/config/utility.cc:96
#8  0x0000555559d9cb73 in Envoy::Config::SubscriptionFactoryImpl::subscriptionFromConfigSource (this=0x55555c204260, config=..., type_url=..., scope=..., callbacks=...) at external/envoy/source/common/config/subscription_factory_impl.cc:33
#9  0x0000555559a06175 in Envoy::Secret::SdsApi::initialize (this=0x55555bec5390) at external/envoy/source/common/secret/sds_api.cc:117
#10 0x0000555559a06c98 in Envoy::Secret::SdsApi::SdsApi(envoy::config::core::v3::ConfigSource, absl::string_view, Envoy::Config::SubscriptionFactory&, Envoy::TimeSource&, Envoy::ProtobufMessage::ValidationVisitor&, Envoy::Stats::Store&, Envoy::Init::Manager&, std::function<void ()>, Envoy::Event::Dispatcher&, Envoy::Api::Api&)::$_0::operator()() const (this=0x55555c43b178) at external/envoy/source/common/secret/sds_api.cc:25
#11 0x0000555559a06b5d in std::_Function_handler<void (), Envoy::Secret::SdsApi::SdsApi(envoy::config::core::v3::ConfigSource, absl::string_view, Envoy::Config::SubscriptionFactory&, Envoy::TimeSource&, Envoy::ProtobufMessage::ValidationVisitor&, Envoy::Stats::Store&, Envoy::Init::Manager&, std::function<void ()>, Envoy::Event::Dispatcher&, Envoy::Api::Api&)::$_0>::_M_invoke(std::_Any_data const&) (__functor=...) at /usr/lib/gcc/x86_64-linux-gnu/7.4.0/../../../../include/c++/7.4.0/bits/std_function.h:316
#12 0x00005555580bf4ee in std::function<void ()>::operator()() const (this=0x55555c43b178) at /usr/lib/gcc/x86_64-linux-gnu/7.4.0/../../../../include/c++/7.4.0/bits/std_function.h:706
#13 0x000055555a0bf590 in Envoy::Init::TargetImpl::TargetImpl(absl::string_view, std::function<void ()>)::$_0::operator()(std::unique_ptr<Envoy::Init::WatcherHandle, std::default_delete<Envoy::Init::WatcherHandle> >) const (this=0x55555c43b170, watcher_handle=...)
    at external/envoy/source/common/init/target_impl.cc:29
#14 0x000055555a0bf3c6 in std::_Function_handler<void (std::unique_ptr<Envoy::Init::WatcherHandle, std::default_delete<Envoy::Init::WatcherHandle> >), Envoy::Init::TargetImpl::TargetImpl(absl::string_view, std::function<void ()>)::$_0>::_M_invoke(std::_Any_data const&, std::unique_ptr<Envoy::Init::WatcherHandle, std::default_delete<Envoy::Init::WatcherHandle> >&&) (__functor=..., __args=...) at /usr/lib/gcc/x86_64-linux-gnu/7.4.0/../../../../include/c++/7.4.0/bits/std_function.h:316
#15 0x000055555a0c024b in std::function<void (std::unique_ptr<Envoy::Init::WatcherHandle, std::default_delete<Envoy::Init::WatcherHandle> >)>::operator()(std::unique_ptr<Envoy::Init::WatcherHandle, std::default_delete<Envoy::Init::WatcherHandle> >) const (this=0x55555c43b1b0, __args=...)
    at /usr/lib/gcc/x86_64-linux-gnu/7.4.0/../../../../include/c++/7.4.0/bits/std_function.h:706
#16 0x000055555a0bdf09 in Envoy::Init::TargetHandleImpl::initialize (this=0x55555c39c300, watcher=...) at external/envoy/source/common/init/target_impl.cc:16
#17 0x0000555559e3d73f in Envoy::Init::ManagerImpl::initialize (this=0x55555c013e60, watcher=...) at external/envoy/source/common/init/manager_impl.cc:55
#18 0x0000555559856820 in Envoy::Server::ListenerImpl::ListenerImpl(envoy::config::listener::v3::Listener const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, Envoy::Server::ListenerManagerImpl&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, bool, bool, unsigned long, unsigned int)::$_1::operator()() const (this=0x55555c41fad8) at external/envoy/source/server/listener_impl.cc:214
#19 0x00005555598566bd in std::_Function_handler<void (), Envoy::Server::ListenerImpl::ListenerImpl(envoy::config::listener::v3::Listener const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, Envoy::Server::ListenerManagerImpl&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, bool, bool, unsigned long, unsigned int)::$_1>::_M_invoke(std::_Any_data const&) (__functor=...) at /usr/lib/gcc/x86_64-linux-gnu/7.4.0/../../../../include/c++/7.4.0/bits/std_function.h:316
#20 0x00005555580bf4ee in std::function<void ()>::operator()() const (this=0x55555c41fad8) at /usr/lib/gcc/x86_64-linux-gnu/7.4.0/../../../../include/c++/7.4.0/bits/std_function.h:706
#21 0x000055555a0bf590 in Envoy::Init::TargetImpl::TargetImpl(absl::string_view, std::function<void ()>)::$_0::operator()(std::unique_ptr<Envoy::Init::WatcherHandle, std::default_delete<Envoy::Init::WatcherHandle> >) const (this=0x55555c41fad0, watcher_handle=...)
    at external/envoy/source/common/init/target_impl.cc:29
#22 0x000055555a0bf3c6 in std::_Function_handler<void (std::unique_ptr<Envoy::Init::WatcherHandle, std::default_delete<Envoy::Init::WatcherHandle> >), Envoy::Init::TargetImpl::TargetImpl(absl::string_view, std::function<void ()>)::$_0>::_M_invoke(std::_Any_data const&, std::unique_ptr<Envoy::Init::WatcherHandle, std::default_delete<Envoy::Init::WatcherHandle> >&&) (__functor=..., __args=...) at /usr/lib/gcc/x86_64-linux-gnu/7.4.0/../../../../include/c++/7.4.0/bits/std_function.h:316
#23 0x000055555a0c024b in std::function<void (std::unique_ptr<Envoy::Init::WatcherHandle, std::default_delete<Envoy::Init::WatcherHandle> >)>::operator()(std::unique_ptr<Envoy::Init::WatcherHandle, std::default_delete<Envoy::Init::WatcherHandle> >) const (this=0x55555c418880, __args=...)
    at /usr/lib/gcc/x86_64-linux-gnu/7.4.0/../../../../include/c++/7.4.0/bits/std_function.h:706
#24 0x000055555a0bdf09 in Envoy::Init::TargetHandleImpl::initialize (this=0x55555c3a4cc0, watcher=...) at external/envoy/source/common/init/target_impl.cc:16
#25 0x0000555559e3d73f in Envoy::Init::ManagerImpl::initialize (this=0x55555beecbd0, watcher=...) at external/envoy/source/common/init/manager_impl.cc:55
#26 0x0000555559821921 in Envoy::Server::RunHelper::RunHelper(Envoy::Server::Instance&, Envoy::Server::Options const&, Envoy::Event::Dispatcher&, Envoy::Upstream::ClusterManager&, Envoy::AccessLog::AccessLogManager&, Envoy::Init::Manager&, Envoy::Server::OverloadManager&, std::function<void ()>)::$_12::operator()() const (this=0x55555c000c60) at external/envoy/source/server/server.cc:567
#27 0x00005555598215ad in std::_Function_handler<void (), Envoy::Server::RunHelper::RunHelper(Envoy::Server::Instance&, Envoy::Server::Options const&, Envoy::Event::Dispatcher&, Envoy::Upstream::ClusterManager&, Envoy::AccessLog::AccessLogManager&, Envoy::Init::Manager&, Envoy::Server::OverloadManager&, std::function<void ()>)::$_12>::_M_invoke(std::_Any_data const&) (__functor=...) at /usr/lib/gcc/x86_64-linux-gnu/7.4.0/../../../../include/c++/7.4.0/bits/std_function.h:316
#28 0x00005555580bf4ee in std::function<void ()>::operator()() const (this=0x55555c204168) at /usr/lib/gcc/x86_64-linux-gnu/7.4.0/../../../../include/c++/7.4.0/bits/std_function.h:706
#29 0x0000555559a5eb60 in Envoy::Upstream::ClusterManagerInitHelper::maybeFinishInitialize (this=0x55555c204138) at external/envoy/source/common/upstream/cluster_manager_impl.cc:173
#30 0x0000555559a5e046 in Envoy::Upstream::ClusterManagerInitHelper::removeCluster (this=0x55555c204138, cluster=...) at external/envoy/source/common/upstream/cluster_manager_impl.cc:106
#31 0x0000555559a5dc63 in Envoy::Upstream::ClusterManagerInitHelper::onClusterInit (this=0x55555c204138, cluster=...) at external/envoy/source/common/upstream/cluster_manager_impl.cc:83
#32 0x0000555559a6c79c in Envoy::Upstream::ClusterManagerInitHelper::addCluster(Envoy::Upstream::Cluster&)::$_0::operator()() const (this=0x7fffffffd9b0) at external/envoy/source/common/upstream/cluster_manager_impl.cc:62
#33 0x0000555559a6c65d in std::_Function_handler<void (), Envoy::Upstream::ClusterManagerInitHelper::addCluster(Envoy::Upstream::Cluster&)::$_0>::_M_invoke(std::_Any_data const&) (__functor=...) at /usr/lib/gcc/x86_64-linux-gnu/7.4.0/../../../../include/c++/7.4.0/bits/std_function.h:316
#34 0x00005555580bf4ee in std::function<void ()>::operator()() const (this=0x7fffffffd9b0) at /usr/lib/gcc/x86_64-linux-gnu/7.4.0/../../../../include/c++/7.4.0/bits/std_function.h:706
#35 0x0000555559d3a576 in Envoy::Upstream::ClusterImplBase::finishInitialization (this=0x55555c0441f0) at external/envoy/source/common/upstream/upstream_impl.cc:999
#36 0x0000555559d42434 in Envoy::Upstream::ClusterImplBase::onInitDone()::$_6::operator()(std::shared_ptr<Envoy::Upstream::Host>, Envoy::Upstream::HealthTransition) const (this=0x55555c20a7f0) at external/envoy/source/common/upstream/upstream_impl.cc:975
#37 0x0000555559d42295 in std::_Function_handler<void (std::shared_ptr<Envoy::Upstream::Host> const&, Envoy::Upstream::HealthTransition), Envoy::Upstream::ClusterImplBase::onInitDone()::$_6>::_M_invoke(std::_Any_data const&, std::shared_ptr<Envoy::Upstream::Host> const&, Envoy::Upstream::HealthTransition&&) (__functor=..., __args=@0x7fffffffdb7c: Envoy::Upstream::HealthTransition::Unchanged, __args=@0x7fffffffdb7c: Envoy::Upstream::HealthTransition::Unchanged) at /usr/lib/gcc/x86_64-linux-gnu/7.4.0/../../../../include/c++/7.4.0/bits/std_function.h:316
#38 0x0000555559e1de7f in std::function<void (std::shared_ptr<Envoy::Upstream::Host> const&, Envoy::Upstream::HealthTransition)>::operator()(std::shared_ptr<Envoy::Upstream::Host> const&, Envoy::Upstream::HealthTransition) const (this=0x55555c20a7f0,
--Type <RET> for more, q to quit, c to continue without paging--
    __args=Envoy::Upstream::HealthTransition::Unchanged, __args=Envoy::Upstream::HealthTransition::Unchanged) at /usr/lib/gcc/x86_64-linux-gnu/7.4.0/../../../../include/c++/7.4.0/bits/std_function.h:706
#39 0x0000555559e19642 in Envoy::Upstream::HealthCheckerImplBase::runCallbacks (this=0x55555bee9690, host=..., changed_state=Envoy::Upstream::HealthTransition::Unchanged) at external/envoy/source/common/upstream/health_checker_base_impl.cc:190
#40 0x0000555559e1a99c in Envoy::Upstream::HealthCheckerImplBase::ActiveHealthCheckSession::setUnhealthy (this=0x55555c03fe00, type=envoy::data::core::v3::NETWORK) at external/envoy/source/common/upstream/health_checker_base_impl.cc:354
#41 0x0000555559e1a9f2 in Envoy::Upstream::HealthCheckerImplBase::ActiveHealthCheckSession::handleFailure (this=0x55555c03fe00, type=envoy::data::core::v3::NETWORK) at external/envoy/source/common/upstream/health_checker_base_impl.cc:360
#42 0x0000555559e1ab3b in Envoy::Upstream::HealthCheckerImplBase::ActiveHealthCheckSession::onTimeoutBase (this=0x55555c03fe00) at external/envoy/source/common/upstream/health_checker_base_impl.cc:391
#43 0x0000555559e1c3c8 in Envoy::Upstream::HealthCheckerImplBase::ActiveHealthCheckSession::ActiveHealthCheckSession(Envoy::Upstream::HealthCheckerImplBase&, std::shared_ptr<Envoy::Upstream::Host>)::$_8::operator()() const (this=0x55555c0b0fa0)
    at external/envoy/source/common/upstream/health_checker_base_impl.cc:236
#44 0x0000555559e1c28d in std::_Function_handler<void (), Envoy::Upstream::HealthCheckerImplBase::ActiveHealthCheckSession::ActiveHealthCheckSession(Envoy::Upstream::HealthCheckerImplBase&, std::shared_ptr<Envoy::Upstream::Host>)::$_8>::_M_invoke(std::_Any_data const&) (__functor=...)
    at /usr/lib/gcc/x86_64-linux-gnu/7.4.0/../../../../include/c++/7.4.0/bits/std_function.h:316
#45 0x00005555580bf4ee in std::function<void ()>::operator()() const (this=0x55555c0b0fa0) at /usr/lib/gcc/x86_64-linux-gnu/7.4.0/../../../../include/c++/7.4.0/bits/std_function.h:706
#46 0x00005555599bfe1d in Envoy::Event::TimerImpl::TimerImpl(Envoy::CSmartPtr<event_base, &event_base_free>&, std::function<void ()>, Envoy::Event::Dispatcher&)::$_0::operator()(int, short, void*) const (this=0xffffffff, arg=0x55555c0b0f20) at external/envoy/source/common/event/timer_impl.cc:15
#47 0x00005555599bfdc6 in Envoy::Event::TimerImpl::TimerImpl(Envoy::CSmartPtr<event_base, &event_base_free>&, std::function<void ()>, Envoy::Event::Dispatcher&)::$_0::__invoke(int, short, void*) (arg=0x55555c0b0f20) at external/envoy/source/common/event/timer_impl.cc:15
#48 0x000055555a344611 in event_process_active_single_queue (base=0x55555bfca000, activeq=0x55555bd7a640, max_to_process=2147483647, endtime=0x0)
    at /build/tmp/_bazel_vscode/59e81b601d532dd328f1d0ef7688287c/sandbox/processwrapper-sandbox/189/execroot/proxyd/external/com_github_libevent_libevent/event.c:1713
#49 0x000055555a33ed0a in event_process_active (base=0x55555bfca000) at /build/tmp/_bazel_vscode/59e81b601d532dd328f1d0ef7688287c/sandbox/processwrapper-sandbox/189/execroot/proxyd/external/com_github_libevent_libevent/event.c:1805
#50 0x000055555a33dbdc in event_base_loop (base=0x55555bfca000, flags=0) at /build/tmp/_bazel_vscode/59e81b601d532dd328f1d0ef7688287c/sandbox/processwrapper-sandbox/189/execroot/proxyd/external/com_github_libevent_libevent/event.c:2047
#51 0x00005555599be228 in Envoy::Event::LibeventScheduler::run (this=0x55555bec6f70, mode=Envoy::Event::Dispatcher::RunType::Block) at external/envoy/source/common/event/libevent_scheduler.cc:47
#52 0x000055555998d45a in Envoy::Event::DispatcherImpl::run (this=0x55555bec6f00, type=Envoy::Event::Dispatcher::RunType::Block) at external/envoy/source/common/event/dispatcher_impl.cc:203
#53 0x000055555981da6b in Envoy::Server::InstanceImpl::run (this=0x55555bdf0000) at external/envoy/source/server/server.cc:592
#54 0x0000555558043e78 in Envoy::MainCommonBase::run (this=0x55555bdab5e8) at external/envoy/source/exe/main_common.cc:112
#55 0x000055555804130e in Envoy::MainCommon::run (this=0x55555bdab000) at bazel-out/k8-dbg/bin/external/envoy/source/exe/_virtual_includes/envoy_main_common_lib/exe/main_common.h:97
#56 0x0000555558040cee in main (argc=3, argv=0x7fffffffe448) at external/envoy/source/exe/main.cc:39
```

Seen on https://github.com/envoyproxy/envoy/commit/8951b3f7b26cbdf21c734d4372555188a42aad65

### PR DESCRIPTION
**Pull request by htuch**
Similar root cause to #4377 - we can't throw exceptions in initialize().

Fixes #10976

Risk level: Low
Testing: Unit/integration test added.

Signed-off-by: Harvey Tuch <htuch@google.com>

### CODE REVIEW

File: test/common/secret/sds_api_test.cc [hunk 1]
```diff
@@ -63,6 +64,23 @@ TEST_F(SdsApiTest, BasicTest) {
   initialize();
 }
 
+// Validate that SdsApi object is created and initialized successfully.
```
Comment (snowp) [inline test/common/secret/sds_api_test.cc:67]: comment seems off since constructions actually fails

**Comment by htuch** [comment]
@snowp turns out there is a slight problem in moving the subscription config to the constructor. Check out this config fragment from one of the integration tests:
```
static_resources:
  clusters:
    - name: cluster_0
      connect_timeout: 5s
      transport_socket:
        name: envoy.transport_sockets.tls
        typed_config:
          "@type": type.googleapis.com/envoy.extensions.transport_sockets.tls.v3.UpstreamTlsContext
          common_tls_context:
            tls_certificates:
              - certificate_chain:
                  filename: /b/f/w/bazel-out/k8-fastbuild/bin/test/integration/sds_dynamic_integration_test.runfiles/envoy/test/config/integration/certs/clientcert.pem
                private_key:
                  filename: /b/f/w/bazel-out/k8-fastbuild/bin/test/integration/sds_dynamic_integration_test.runfiles/envoy/test/config/integration/certs/clientkey.pem
            validation_context_sds_secret_config:
              name: validation_secret
              sds_config:
                api_config_source:
                  api_type: GRPC
                  grpc_services:
                    - envoy_grpc:
                        cluster_name: sds_cluster
      load_assignment:
        cluster_name: cluster_0
        endpoints:
          - lb_endpoints:
              - endpoint:
                  address:
                    socket_address:
                      address: 127.0.0.1
                      port_value: 36887
    - name: sds_cluster
      connect_timeout: 5s
      http2_protocol_options:
        {}
      load_assignment:
        cluster_name: cluster_0
        endpoints:
          - lb_endpoints:
              - endpoint:
                  address:
                    socket_address:
                      address: 127.0.0.1
                      port_value: 42657
```

`cluster_0` refers to `sds_cluster` via its transport socket, but we see an error:
```
[2020-05-18 00:30:06.019][24][critical][main] [source/server/server.cc:99] error initializing configuration '/b/f/w/_tmp/d0a45f39a3cb27a1acdbf12470d6dce1/bootstrap.json': envoy.config.core.v3.ApiConfigSource must have a statically defined non-EDS cluster: 'sds_cluster' does not exist, was added via api, or is an EDS cluster
```
because `sds_cluster` is defined after `cluster_0`, with the check occurring before `sds_cluster` is processed.

I think the "cleanest" solution here might be to have `ClusterManager` be aware of the set of pending config fragments that are being considered for loading in the current stack frame, e.g. `pendingClusters()`, and have https://github.com/envoyproxy/envoy/blob/cf2d4df8981bcff50d280a040cfa38a60b48f27f/source/common/config/utility.cc#L139 be able to query this. WDYT?

**Comment by snowp** [comment]
I like that idea, hopefully it will future proof us against other new uses cases where we need to reference a static cluster.

**Comment by htuch** [comment]
@snowp this is now updated, PTAL.

**Pull request merged**